### PR TITLE
Fixed #30644 -- Made introspection use pg_table_is_visible() instead of filtering by public schema on PostgreSQL.

### PR DESCRIPTION
**Pull request by georgi-yanchev**
Only some of the functions in `db/backends/postgresql/introspection.py` respect visible tables. The remaining function either assume `public` (hardcoded) or return everything found mixed together.

This PR updates the remaining functions to only deal with tables visible in the current `search_path` (Postgres schema).

This partially addresses https://code.djangoproject.com/ticket/22673

### CODE REVIEW

**Comment by georgi-yanchev** [comment]
Full description of the problem at: https://code.djangoproject.com/ticket/30644

File: django/db/backends/postgresql/introspection.py [hunk 1]
```diff
@@ -138,7 +139,15 @@ def get_key_columns(self, cursor, table_name):
...
-            WHERE kcu.table_name = %s AND tc.constraint_type = 'FOREIGN KEY'
+            JOIN pg_catalog.pg_class cl
+                ON cl.relname = kcu.table_name
+            JOIN pg_catalog.pg_namespace n
+                ON cl.relnamespace = n.oid
+                    AND n.nspname = kcu.table_schema
+            WHERE kcu.table_name = %s
+                AND tc.constraint_type = 'FOREIGN KEY'
+                AND cl.relkind IN ('r', 'v')
+                AND pg_catalog.pg_table_is_visible(cl.oid)
```
Comment (felixxm) [inline django/db/backends/postgresql/introspection.py]: Is all of this necessary?, why not just:
```SQL
JOIN pg_class c ON c.relname = kcu.table_name
WHERE
    kcu.table_name = %s AND
    tc.constraint_type = 'FOREIGN KEY' AND
    pg_catalog.pg_table_is_visible(c.oid)
```
You don't need to add `pg_catalog.`

File: django/db/backends/postgresql/introspection.py [hunk 2]
```diff
@@ -212,6 +221,8 @@ def get_constraints(self, cursor, table_name):
                 LEFT JOIN pg_am am ON c2.relam = am.oid
                 LEFT JOIN pg_attribute attr ON attr.attrelid = c.oid AND attr.attnum = idx.key
                 WHERE c.relname = %s
+                    AND c.relnamespace = c2.relnamespace
```
Comment (felixxm) [inline django/db/backends/postgresql/introspection.py]: It looks unnecessary to me.

File: django/db/backends/postgresql/introspection.py [hunk 3]
```diff
@@ -123,6 +122,8 @@ def get_relations(self, cursor, table_name):
             LEFT JOIN pg_attribute a1 ON c1.oid = a1.attrelid AND a1.attnum = con.conkey[1]
             LEFT JOIN pg_attribute a2 ON c2.oid = a2.attrelid AND a2.attnum = con.confkey[1]
             WHERE c1.relname = %s AND con.contype = 'f'
+                AND c1.relnamespace = c2.relnamespace
```
Comment (felixxm) [inline django/db/backends/postgresql/introspection.py]: Is this necessary?

File: django/db/backends/postgresql/introspection.py (hunk 1 above)
Comment (georgi-yanchev) [reply django/db/backends/postgresql/introspection.py]: 1. Without INNER JOIN of `pg_catalog` we are not filtering by schema name.

2. I'm removing `pg_catalog.` prefix.

3. It seems that the query from `get_relations` returns the same data. I was thinking of reusing it, but it seemed too intrusive for this PR.

**Comment by georgi-yanchev** [comment]
I'm testing these changes by typing into my django shell. Creating automated tests sounds challenging.

**Comment by schinckel** [comment]
Thanks for this PR. The hardcoding of "public" schema was the rationalé behind my overriding this function in a project I wrote: https://bitbucket.org/schinckel/django-boardinghouse/src/8a2320d4be97481bafec79d8f27f90d2638ac2d0/boardinghouse/backends/postgres/schema.py#lines-15:106

I used a totally different approach (explicitly using the schema names that were passed in), but I think the `pg_catalog.pg_table_is_visible()` function is much nicer.

(I also combined all of these queries into one for my solution, however that was because this function was called once per tenant schema, and there could be thousands of them. Ultimately, having to apply migrations to each schema was the reason I stopped developing that project, and don't use it in production).

It does have a bunch of tests that shows how you can potentially test the `get_constraints()` function: possibly that can help you with some automated tests on this fix?

**Comment by schinckel** [comment]
Ugh. Forgot the link to tests: they are in https://bitbucket.org/schinckel/django-boardinghouse/src/8a2320d4be97481bafec79d8f27f90d2638ac2d0/tests/tests/test_migrations.py?at=default#lines-40

**Comment by georgi-yanchev** [comment]
Thanks. 

@felixxm What else is needed to get this merged?

File: django/db/backends/postgresql/introspection.py (hunk 1 above)
Comment (felixxm) [reply django/db/backends/postgresql/introspection.py]: I was confused why we have two different queries in `get_key_columns()` and `get_relations()` that return the same columns but in a different format, so I unified them in #11577. What do you think? it should make this change much simpler.

Comment (georgi-yanchev) [reply django/db/backends/postgresql/introspection.py]: Looks good to me. If you merge it, I can rebase this PR on top.

Comment (felixxm) [reply django/db/backends/postgresql/introspection.py]: Merged.

**Comment by georgi-yanchev** [comment]
Merged upstream changes to fix conflicts.

@felixxm Do you know why the tests fail so often in a rather random fashion? Is there a way for me to retry failed test(s) without pushing new commits?

Thank you.

**Comment by felixxm** [comment]
Tests failures are mostly related with force push, see also [Troubleshooting](https://code.djangoproject.com/wiki/Jenkins#Troubleshooting).

**Comment by felixxm** [comment]
@georgi-yanchev Thanks :+1: Welcome aboard :boat:

**Comment by georgi-yanchev** [comment]
Thank you so much. Any chance to get it merged into Django 2 as well?

**Comment by felixxm** [comment]
Unfortunately not, this doesn't qualify for a backport.

**Pull request merged**